File: memoization/countcons.py
def countConstruct(target,words,memo={}):
    totalCount = 0
    if target in memo:
        return memo[target]
    if target == '':
        return 1
    for w in words:
        try:
            if target.index(w) == 0:
                prefix = target[len(w):]
                totalCount += countConstruct(prefix,words,memo)
        except:
            pass
    memo[target] = totalCount
    return totalCount

print(countConstruct('abcdef',['ab','abc','abcd','def','cd']))
print(countConstruct('skateboard',['s','k','t','ate','bo','boar']))
print(countConstruct('eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef',['e','ee','eee','eeee','eeeee','eeeeee']))

File: tabulation/countcons.js
const countConstruct = (target,words) => {
    table = Array(target.length + 1).fill(0);
    table[0] = 1;
    for(let i=0;i<=target.length;i++){
        if(table[i]!=0){
            for(let w of words){
                if(target.slice(i,i+w.length) === w){
                    table[i+w.length] += table[i];
                }
            }
        }
    }
    return table[target.length];
}

console.log(countConstruct('abcdef',['ab','abc','abcd','def','cd','ef']))
console.log(countConstruct('skateboard',['s','k','t','ate','bo','boar']))
console.log(countConstruct('eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef',['ef','ee','eee','eeee','eeeee','eeeeee']))
